{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a17a866f",
   "metadata": {},
   "source": [
    "# NCBI Datasets (07/21/2022)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26f69883",
   "metadata": {},
   "source": [
    "### Table of contents\n",
    "* [Part I: Introduction to NCBI Datasets](#Part-I) \n",
    "* [Part II: Getting a list of gene-ids per species](#Part-II)\n",
    "* [Part III: Getting ortholog data packages from a list of gene-ids](#Part-III)\n",
    "* [Part IV: Closer look at the ortholog metadata files](#Part-IV)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d60a05cb",
   "metadata": {},
   "source": [
    "## Part I: Introduction to NCBI Datasets<a class=\"anchor\" id=\"Part-I\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b785d5b",
   "metadata": {},
   "source": [
    "### How is `datasets` organized?\n",
    "\n",
    "[NCBI datasets](https://www.ncbi.nlm.nih.gov/datasets/docs/v1/download-and-install/) is a command line tool that allows users to download data packages (data + metadata) or look at metadata summaries for genomes, RefSeq annotated genes, curated ortholog sets and SARS-Cov-2 virus sequences and proteins. The program follows a hierarchy that makes it easier for users to select exact which options they would like to use. In addition to the program commands, additional flags are available for filtering the results. We will go over those during this tutorial.\n",
    "<img src=\"https://www.ncbi.nlm.nih.gov/datasets/docs/v1/datasets_schema_complete.svg\" alt=\"datasets\" style=\"width: 800px;\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54d048ca",
   "metadata": {},
   "source": [
    "In addition to `datasets`, we will be using `jq` (JSON parser) and `dataformat`, a `datasets`companion tool, to take a look at the metadata information. Our metadata reports are almost all in JSON or [JSON Lines](https://jsonlines.org/) format. We put together a [jq cheat sheet]( https://github.com/ncbi/datasets/blob/workshop-cshl-2021/training/cshl-2021/jq_cheatsheet.md) to help you extract information from those files.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7f190f7",
   "metadata": {},
   "source": [
    "For example: if I want to download the cat reference genome (<i>Felis catus</i>), I would use the command below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2e1c62b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Process is interrupted.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#Download the cat reference genome, with associated annotation files and metadata\n",
    "\n",
    "datasets download genome taxon cat --reference --no-progressbar\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c7bdc70",
   "metadata": {},
   "source": [
    "Instead of downloading a data package, I could instead look at the metadata information by using the `summary`command. Here, I'm pipping it to [`jq`](https://stedolan.github.io/jq/) so it's easier to read:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "42a87f34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"assemblies\": [\n",
      "    {\n",
      "      \"assembly\": {\n",
      "        \"annotation_metadata\": {\n",
      "          \"busco\": {\n",
      "            \"busco_lineage\": \"carnivora_odb10\",\n",
      "            \"busco_ver\": \"4.1.4\",\n",
      "            \"complete\": 0.9884154,\n",
      "            \"duplicated\": 0.0073093367,\n",
      "            \"fragmented\": 0.0021376363,\n",
      "            \"missing\": 0.009446973,\n",
      "            \"single_copy\": 0.98110604,\n",
      "            \"total_count\": \"14502\"\n",
      "          },\n",
      "          \"file\": [\n",
      "            {\n",
      "              \"estimated_size\": \"26355419\",\n",
      "              \"type\": \"GENOME_GFF\"\n",
      "            },\n",
      "            {\n",
      "              \"estimated_size\": \"1026987832\",\n",
      "              \"type\": \"GENOME_GBFF\"\n",
      "            },\n",
      "            {\n",
      "              \"estimated_size\": \"68647008\",\n",
      "              \"type\": \"RNA_FASTA\"\n",
      "            },\n",
      "            {\n",
      "              \"estimated_size\": \"16615960\",\n",
      "              \"type\": \"PROT_FASTA\"\n",
      "            },\n",
      "            {\n",
      "              \"estimated_size\": \"25749352\",\n",
      "              \"type\": \"GENOME_GTF\"\n",
      "            },\n",
      "            {\n",
      "              \"estimated_size\": \"21917617\",\n",
      "              \"type\": \"CDS_FASTA\"\n",
      "            }\n",
      "          ],\n",
      "          \"name\": \"NCBI Annotation Release 105\",\n",
      "          \"release_date\": \"2021-10-27\",\n",
      "          \"release_number\": \"105\",\n",
      "          \"report_url\": \"https://www.ncbi.nlm.nih.gov/genome/annotation_euk/Felis_catus/105/\",\n",
      "          \"source\": \"NCBI RefSeq\",\n",
      "          \"stats\": {\n",
      "            \"gene_counts\": {\n",
      "              \"protein_coding\": 20453,\n",
      "              \"total\": 39307\n",
      "            }\n",
      "          }\n",
      "        },\n",
      "        \"assembly_accession\": \"GCF_018350175.1\",\n",
      "        \"assembly_category\": \"representative genome\",\n",
      "        \"assembly_level\": \"Chromosome\",\n",
      "        \"bioproject_lineages\": [\n",
      "          {\n",
      "            \"bioprojects\": [\n",
      "              {\n",
      "                \"accession\": \"PRJNA684600\",\n",
      "                \"parent_accessions\": [\n",
      "                  \"PRJNA701031\"\n",
      "                ],\n",
      "                \"title\": \"Felis catus - maternal haplotype of Felis catus x Leopardus geoffroyi (GxD-2)\"\n",
      "              },\n",
      "              {\n",
      "                \"accession\": \"PRJNA701031\",\n",
      "                \"title\": \"Felis catus x Oncifelis geoffroyi (Felis catus) Genome sequencing and assembly\"\n",
      "              }\n",
      "            ]\n",
      "          }\n",
      "        ],\n",
      "        \"biosample\": {\n",
      "          \"accession\": \"SAMN19729387\",\n",
      "          \"attributes\": [\n",
      "            {\n",
      "              \"name\": \"isolate\",\n",
      "              \"value\": \"Fca126\"\n",
      "            },\n",
      "            {\n",
      "              \"name\": \"dev_stage\",\n",
      "              \"value\": \"Adult\"\n",
      "            },\n",
      "            {\n",
      "              \"name\": \"sex\",\n",
      "              \"value\": \"female\"\n",
      "            },\n",
      "            {\n",
      "              \"name\": \"cell_type\",\n",
      "              \"value\": \"Fibroblast\"\n",
      "            },\n",
      "            {\n",
      "              \"name\": \"tissue\",\n",
      "              \"value\": \"Fibroblast\"\n",
      "            }\n",
      "          ],\n",
      "          \"bioprojects\": [\n",
      "            {\n",
      "              \"accession\": \"PRJNA684600\"\n",
      "            }\n",
      "          ],\n",
      "          \"description\": {\n",
      "            \"organism\": {\n",
      "              \"organism_name\": \"Felis catus\",\n",
      "              \"tax_id\": 9685\n",
      "            },\n",
      "            \"title\": \"Domestic Cat\"\n",
      "          },\n",
      "          \"last_updated\": \"2021-10-15T15:05:54.236\",\n",
      "          \"models\": [\n",
      "            \"Model organism or animal\"\n",
      "          ],\n",
      "          \"owner\": {\n",
      "            \"contacts\": [\n",
      "              {}\n",
      "            ],\n",
      "            \"name\": \"Texas A&M University\"\n",
      "          },\n",
      "          \"package\": \"Model.organism.animal.1.0\",\n",
      "          \"publication_date\": \"2021-06-23T00:00:00.000\",\n",
      "          \"sample_ids\": [\n",
      "            {\n",
      "              \"label\": \"Sample name\",\n",
      "              \"value\": \"Fca126\"\n",
      "            }\n",
      "          ],\n",
      "          \"status\": {\n",
      "            \"status\": \"live\",\n",
      "            \"when\": \"2021-06-23T00:50:09.247\"\n",
      "          },\n",
      "          \"submission_date\": \"2021-06-16T11:19:05.613\"\n",
      "        },\n",
      "        \"biosample_accession\": \"SAMN19729387\",\n",
      "        \"blast_url\": \"https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&PROG_DEF=blastn&BLAST_SPEC=GDH_GCF_018350175.1\",\n",
      "        \"chromosomes\": [\n",
      "          {\n",
      "            \"accession_version\": \"NC_058368.1\",\n",
      "            \"gc_count\": \"95626933\",\n",
      "            \"length\": \"239367248\",\n",
      "            \"name\": \"A1\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058369.1\",\n",
      "            \"gc_count\": \"71807841\",\n",
      "            \"length\": \"169388855\",\n",
      "            \"name\": \"A2\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058370.1\",\n",
      "            \"gc_count\": \"60458931\",\n",
      "            \"length\": \"140443288\",\n",
      "            \"name\": \"A3\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058371.1\",\n",
      "            \"gc_count\": \"80491461\",\n",
      "            \"length\": \"205367284\",\n",
      "            \"name\": \"B1\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058372.1\",\n",
      "            \"gc_count\": \"61268826\",\n",
      "            \"length\": \"151959158\",\n",
      "            \"name\": \"B2\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058373.1\",\n",
      "            \"gc_count\": \"62394238\",\n",
      "            \"length\": \"148491486\",\n",
      "            \"name\": \"B3\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058374.1\",\n",
      "            \"gc_count\": \"58984987\",\n",
      "            \"length\": \"142168536\",\n",
      "            \"name\": \"B4\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058375.1\",\n",
      "            \"gc_count\": \"92215633\",\n",
      "            \"length\": \"221611373\",\n",
      "            \"name\": \"C1\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058376.1\",\n",
      "            \"gc_count\": \"63151821\",\n",
      "            \"length\": \"158578461\",\n",
      "            \"name\": \"C2\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058377.1\",\n",
      "            \"gc_count\": \"49274970\",\n",
      "            \"length\": \"115366950\",\n",
      "            \"name\": \"D1\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058378.1\",\n",
      "            \"gc_count\": \"38272432\",\n",
      "            \"length\": \"88083857\",\n",
      "            \"name\": \"D2\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058379.1\",\n",
      "            \"gc_count\": \"40960342\",\n",
      "            \"length\": \"94435393\",\n",
      "            \"name\": \"D3\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058380.1\",\n",
      "            \"gc_count\": \"40890327\",\n",
      "            \"length\": \"95154158\",\n",
      "            \"name\": \"D4\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058381.1\",\n",
      "            \"gc_count\": \"29254025\",\n",
      "            \"length\": \"61876196\",\n",
      "            \"name\": \"E1\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058382.1\",\n",
      "            \"gc_count\": \"28646901\",\n",
      "            \"length\": \"61988844\",\n",
      "            \"name\": \"E2\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058383.1\",\n",
      "            \"gc_count\": \"20036117\",\n",
      "            \"length\": \"41437797\",\n",
      "            \"name\": \"E3\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058384.1\",\n",
      "            \"gc_count\": \"29723076\",\n",
      "            \"length\": \"69239673\",\n",
      "            \"name\": \"F1\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058385.1\",\n",
      "            \"gc_count\": \"34034866\",\n",
      "            \"length\": \"83466477\",\n",
      "            \"name\": \"F2\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_058386.1\",\n",
      "            \"gc_count\": \"52118309\",\n",
      "            \"length\": \"127842892\",\n",
      "            \"name\": \"X\"\n",
      "          },\n",
      "          {\n",
      "            \"accession_version\": \"NC_001700.1\",\n",
      "            \"gc_count\": \"6860\",\n",
      "            \"length\": \"17009\",\n",
      "            \"name\": \"MT\"\n",
      "          },\n",
      "          {\n",
      "            \"gc_count\": \"4687119\",\n",
      "            \"length\": \"9462103\",\n",
      "            \"name\": \"Un\"\n",
      "          }\n",
      "        ],\n",
      "        \"contig_n50\": 90731473,\n",
      "        \"display_name\": \"F.catus_Fca126_mat1.0\",\n",
      "        \"estimated_size\": \"2723004839\",\n",
      "        \"gc_count\": \"1014299155\",\n",
      "        \"org\": {\n",
      "          \"assembly_counts\": {\n",
      "            \"node\": 6,\n",
      "            \"subtree\": 6\n",
      "          },\n",
      "          \"common_name\": \"domestic cat\",\n",
      "          \"isolate\": \"Fca126\",\n",
      "          \"key\": \"9685\",\n",
      "          \"merged_tax_ids\": [\n",
      "            \"36475\"\n",
      "          ],\n",
      "          \"parent_tax_id\": \"9682\",\n",
      "          \"rank\": \"SPECIES\",\n",
      "          \"sci_name\": \"Felis catus\",\n",
      "          \"sex\": \"female\",\n",
      "          \"tax_id\": \"9685\",\n",
      "          \"title\": \"domestic cat\"\n",
      "        },\n",
      "        \"paired_assembly_accession\": \"GCA_018350175.1\",\n",
      "        \"seq_length\": \"2425730029\",\n",
      "        \"submission_date\": \"2021-05-13\",\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        \"submitter\": \"Texas A&M University\"\n",
      "      }\n",
      "    }\n",
      "  ],\n",
      "  \"total_count\": 1\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#Check the metadata information for the cat reference genome\n",
    "\n",
    "datasets summary genome taxon cat --reference | jq ."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd1ac2d9",
   "metadata": {},
   "source": [
    "### Data packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50d70bc5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d66b33f7",
   "metadata": {},
   "source": [
    "NCBI Datasets delivers data as <u>data packages</u>, which which are zip archives containing both data (FASTA, GFF3, GTF, GBFF) and metadata files (JSON, JSON-Lines). The image below shows the contents of all data packages. Files are included depending on their availability. For example: for an annotated genome, the data package would include FASTA files (genomic, transcript, protein and CDS sequences) and annotation files (GFF3, GTF and GBFF).\n",
    "\n",
    "\n",
    "<img src=\"./images/datapackages.png\" alt=\"data_package\" style=\"width: 800px;\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed84aedf",
   "metadata": {},
   "source": [
    "### How to get help when using the command line\n",
    "\n",
    "Since `datasets` is a very hierarchical program, we can use that characteristic to our advantage to get very specific help.   For example: if we type `datasets --help`, we will see the first level of commands available.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d219e750",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "datasets is a command-line tool that is used to query and download biological sequence data\n",
      "across all domains of life from NCBI databases.\n",
      "\n",
      "Refer to NCBI's [download and install](https://www.ncbi.nlm.nih.gov/datasets/docs/download-and-install/) documentation for information about getting started with the command-line tools.\n",
      "\n",
      "Usage\n",
      "  datasets [command]\n",
      "\n",
      "Data Retrieval Commands\n",
      "  summary              print a summary of a gene or genome dataset\n",
      "  download             download a gene, genome or coronavirus dataset as a zip file\n",
      "  rehydrate            rehydrate a downloaded, dehydrated dataset\n",
      "\n",
      "Miscellaneous Commands\n",
      "  completion           generate autocompletion scripts\n",
      "  version              print the version of this client and exit\n",
      "  help                 Help about any command\n",
      "\n",
      "Flags\n",
      "      --api-key string   NCBI Datasets API Key\n",
      "  -h, --help             help for datasets\n",
      "      --no-progressbar   hide progress bar\n",
      "\n",
      "Use datasets help <command> for detailed help about a command.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "datasets --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db05f935",
   "metadata": {},
   "source": [
    "Notice the difference from when we type `datasets summary genome taxon formicidae --help`  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a09930dd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Print a summary of a genome dataset by taxon (NCBI Taxonomy ID, scientific or common name at any tax rank). The summary is returned in JSON format.\n",
      "\n",
      "Refer to NCBI's [download and install](https://www.ncbi.nlm.nih.gov/datasets/docs/download-and-install/) documentation for information about getting started with the command-line tools.\n",
      "\n",
      "Usage\n",
      "  datasets summary genome taxon [flags]\n",
      "\n",
      "Examples\n",
      "  datasets summary genome taxon human\n",
      "  datasets summary genome taxon \"mus musculus\"\n",
      "  datasets summary genome taxon 10116\n",
      "\n",
      "Flags\n",
      "  -h, --help              help for taxon\n",
      "      --tax-exact-match   exclude sub-species when a species-level taxon is specified\n",
      "\n",
      "\n",
      "Global Flags\n",
      "  -a, --annotated                only include genomes with annotation\n",
      "      --api-key string           NCBI Datasets API Key\n",
      "      --as-json-lines            Stream results as newline delimited JSON-Lines\n",
      "      --assembly-level string    restrict assemblies to a comma-separated list of one or more of: chromosome, complete_genome, contig, scaffold\n",
      "      --assembly-source string   restrict assemblies to refseq or genbank only\n",
      "      --assmaccs                 return only assembly accessions\n",
      "      --limit string             limit the number of genome summaries returned\n",
      "                                 * \"all\": returns all matching genome summaries\n",
      "                                 * \"none\": returns only the number of matching genome summaries\n",
      "                                 * a number: returns the specified number of matching genome summaries\n",
      "                                  (default \"all\")\n",
      "      --no-progressbar           hide progress bar\n",
      "      --reference                limit to reference and representative (GCF_ and GCA_) assemblies\n",
      "      --released-before string   only include genomes that have been released before a specified date (MM/DD/YYYY)\n",
      "      --released-since string    only include genomes that have been released after a specified date (MM/DD/YYYY)\n",
      "      --search strings           only include genomes that have the specified text in the\n",
      "                                 searchable fields: species and infraspecies, assembly name and submitter\n",
      "                                 To provide multiple strings '--search' can be included multiple times\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "datasets summary genome taxon formicidae --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92c031a1",
   "metadata": {},
   "source": [
    "## Part II: Getting a list of gene-ids per species <a class=\"anchor\" id=\"Part-II\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b4d055b",
   "metadata": {},
   "source": [
    "Independent of choosing `datasets download` or `datasets summary`, there are three options for retrieving gene information:\n",
    "- accession\n",
    "- gene-id\n",
    "- symbol\n",
    "\n",
    "When choosing any of those three options, you will retrieve the gene information for the **reference** taxon. Like this:\n",
    "\n",
    "`datasets download gene accession XR_002738142.1`  \n",
    "`datasets download gene gene-id 101081937`  \n",
    "`datasets download gene symbol BRCA1 --taxon cat`  \n",
    "\n",
    "All three commands will download the same gene from the cat (<i>Felis catus</i>) <u>reference genome</u>. \n",
    "\n",
    "#### accession\n",
    "Unique identifier. Accession includes RefSeq RNA and protein accessions. Since it's unique, taxon is implied (aka there will never be two sequences from different taxa with the same accession number).  \n",
    "\n",
    "#### gene-id\n",
    "Also a unique identifier. Every RefSeq genome annotated has a unique set of identifiers. For example: the gene-id for BRCA1 in human is 672, while in cat is 101081937.  \n",
    "\n",
    "#### symbol\n",
    "Differently from accession and gene-id, gene symbol is not unique and means different things in different taxonomic groups. If using the symbol option, you should specify the species. The default option is human.\n",
    "\n",
    "**Remember**: both `summary` and `download` will return results for the **reference assembly** of a <u>single species</u>. If you want to download a curated set of the same gene for multiple taxa, you should use the option `ortholog`. We'll talk more about it later. For reference, here's the JSON organization of the gene summary metadata. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b8c4f1b",
   "metadata": {},
   "source": [
    "### Now let's take a look at a gene example:\n",
    "\n",
    "<img src=\"./images/gene_json_gene-id.png\" />"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5196c345",
   "metadata": {},
   "source": [
    "We want to extract a list of gene-ids for *Drosophila melanogaster* (tax-id ). We will use that list later to download ortholog sets for each gene-id. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e2b6fe8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Get list of gene-ids for D. melanogaster and save as a txt file\n",
    "datasets summary gene taxon 7227 --as-json-lines | jq -r '.gene.gene_id' > dmel_gene-ids.txt\n",
    "\n",
    "#alternative\n",
    "# datasets summary gene taxon 7227 | jq -r '.genes[].gene.gene_id' > dmel_gene-ids.txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e748073a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Count the number of lines (genes) in the list\n",
    "wc -l dmel_gene-ids.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c908cc8b",
   "metadata": {},
   "source": [
    "## Part V: Accessing orthologs <a class=\"anchor\" id=\"Part-V\"></a>\n",
    "\n",
    "### Orthologs\n",
    "\n",
    "The options to retrieve ortholog sets are the same as those for genes. We'll go over the differences when using each option:\n",
    "\n",
    "- accession\n",
    "- gene-id\n",
    "- symbol\n",
    "\n",
    "When choosing any of those three options, you will download the **full ortholog set** to which they belong (unless you use additional filtering. We'll cover it below). Like this:\n",
    "\n",
    "`datasets download ortholog accession XR_002738142.1`  \n",
    "`datasets download ortholog gene-id 101081937`  \n",
    "`datasets download ortholog symbol BRCA1 --taxon cat`  \n",
    "\n",
    "All three commands will download the **same** ortholog set. \n",
    "\n",
    "---\n",
    "\n",
    "#### <font color='blue'>Wait, but what is an ortholog set?</font>\n",
    "\n",
    ">An ortholog set, or ortholog gene group, is a group of sequences that have been identified by the NCBI genome annotation team as homologous genes related to each other by speciation events. They are identified by a combination of protein similarity + local syntheny information. \n",
    "Currently, NCBI has ortholog sets calculated for vertebrates and some insects. \n",
    "\n",
    "\n",
    "#### accession\n",
    "Unique identifier. Accession includes RefSeq accession RNA and protein sequences. Since it's unique, taxon is implied (aka there will never be two sequences from different taxa with the same accession number).\n",
    "\n",
    "#### gene-id\n",
    "Also an unique identifier. Every RefSeq genome annotated has a unique set of identifiers. For example: the gene-id for BRCA1 in human is 672, while in cat is 101081937. You can use either one (672 or 101081937) to get the same vertebrate BRCA1 ortholog set.\n",
    "\n",
    "#### symbol\n",
    "Differently from accession and gene-id, gene symbol is not unique and means different things in different taxonomic groups. For example: the P53 ortholog set in vertebrates is different from the insect set. If using the symbol option, you should specify the taxonomic group. The default option is human. Note that if you want ortholog sets from multiple vertebrate species, you might end up downloading the same ortholog set multiple times. Like this: \n",
    "\n",
    "`datasets download ortholog symbol brca1 --taxon cat`  \n",
    "`datasets download ortholog symbol brca1 --taxon chicken`  \n",
    "`datasets download ortholog symbol brca1 --taxon \"chelonia mydas\"`  \n",
    "\n",
    "If that's the case, how do you filter the ortholog set to include *only* your taxonomic group of interest?\n",
    "\n",
    "### Applying a taxonomic filter to the ortholog set\n",
    "\n",
    "For the orthologs, `datasets` provides the flag `--taxon-filter`, which allows the user to restrict the summary or download to one or multiple taxonomic groups.  `--taxon` and `--taxon-filter` have different effects on the data package/summary output. A few examples:\n",
    "\n",
    "- `datasets summary ortholog symbol brca1 --taxon-filter \"felis catus\"`  \n",
    "Prints a json metadata summary of the gene brca1 for the domestic cat. \n",
    "We did not specify a `--taxon` because the default is human, and Felidae and human are part of the same brca1 ortholog set.   \n",
    "\n",
    "  \n",
    "\n",
    "- `datasets summary ortholog symbol brca1 --taxon \"felis catus\"`  \n",
    "Even though this option looks almost the same as the one above, the result is *very different*. Here, we're asking `datasets` to find the ortholog set to which the gene brca1 in the domestic cat belongs. And `datasets` will download the <u>entire</u> ortholog set, not only the sequences for the domestic cat.\n",
    "\n",
    "\n",
    "- `datasets summary ortholog symbol brca1 --taxon \"felis catus\" --taxon-filter \"felis catus\"`  \n",
    "gives you the same result as `datasets summary ortholog symbol brca1 --taxon-filter \"felis catus\"`\n",
    "\n",
    "\n",
    "The summary metadata for orthologs is presented in JSON Lines, which means that each gene entry is in a different line. Here's the diagram to help you create queries.\n",
    "  \n",
    "<img src=\"./images/ortholog_jsonl.drawio.png\" />"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3673c066",
   "metadata": {},
   "source": [
    "#### We are going to do the following steps:\n",
    "- Use a loop to download an ortholog data package for each gene-id in the list\n",
    "- unzip it to the folder ortholog\n",
    "- look at the files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4040acda",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# download an ortholog data package per gene-id. We will be using a reduced set (`head -n20`) as example:\n",
    "\n",
    "mkdir orthologs;\n",
    "head -n20 dmel_gene-ids.txt | while read GENEID; do \n",
    "        echo ${GENEID}; \n",
    "        datasets download ortholog gene-id \"${GENEID}\" \n",
    "        --filename ./orthologs/$GENEID.zip \n",
    "        --taxon-filter 7215; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c295a289",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Check the number of ortholog sets downloaded:\n",
    "ls orthologs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "299dc9d9",
   "metadata": {},
   "source": [
    "## Part II: Accessing metadata<a class=\"anchor\" id=\"Part-II\"></a>\n",
    "\n",
    "First, let's extract some metadata information from each of those data packages using `dataformat`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd676b9b",
   "metadata": {},
   "source": [
    "`dataformat`is a `datasets`companion tool and it allows you to export metadata information to tabular or Excel format. Here, we will use `dataformat` to extract the following info from each ortholog data package: organism, accession, CDS range start and CDS range end. You can find more info about the JSON schemas in this page #ADD LINK#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22330aa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "dataformat tsv gene \n",
    "    --fields tax-name,transcript-cds-accession,transcript-cds-range-start,transcript-cds-range-stop \n",
    "    --package 10178785.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6ca87c8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
